### DISCUSSION TITLE
Apple Pay / Google Pay

### DISCUSSION THREAD

**Discussion by RitterKnightCreative** [craftcms/commerce-stripe#164] [Ideas]
Stripe supports both but both Apple Pay in particular works around the "standard" checkout process of grabbing data.

Is there a good canonical example of supporting Apple Pay in Craft Commerce?  

[This](https://github.com/aladrach/craft-apple-pay/blob/master/apple-pay-info.twig) got us close, but with the changes to C2, we had to change a few things to make it go and  I was able to get Stripe to charge the card but Commerce is not accepting the order.

**Comment by andris-sevcenko** [comment]
I'll take a stab at this when I get around to issues 26 through to 28.

**Comment by philipboomy** [comment]
@andris-sevcenko Any chance you have had time to look at this? 

Thanks

**Comment by andris-sevcenko** [comment]
I wish :(

**Comment by RitterKnightCreative** [comment]
@philipboomy I actually ended up getting this working, I'll see what I can do about sharing the code but the link to the repo above got us close. 

There was actually an unrelated Commerce bug that got fixed... Commerce was thinking the order had changed but it actually had not...

**Comment by philipboomy** [comment]
Thats great you got it working. If you could share some of the code then I would really appreciate it. No rush though.

**Comment by andris-sevcenko** [comment]
@RitterKnightCreative I'm also interested if you can share the code. If it's not hard to include in the gateway that would lessen the friction for me to add this to the gateway.

**Comment by Tam** [comment]
@RitterKnightCreative We're working on a plugin to add Google/Apple pay support if you're interested in trying it out (not released yet as we're testing it on client sites first).

https://github.com/ethercreative/web-payments

**Comment by peteeveleigh** [comment]
@Tam I'd be interested in trying the plugin but can't seem to get it installed. How should that be done?

Nevermind! I was being dumb. Got it installed and seems to work OK so far!

**Comment by RitterKnightCreative** [comment]
@Tam Nice!  We completed our implementation a while back as well, got sidetracked and sort of forgot about this thread but I'll get our code up here on GH as well and I'll take a look at yours...

What's nice about Stripe's implementation is it works for not only Android but also Chrome's autofill.

**Comment by philipboomy** [comment]
@RitterKnightCreative That would be awesome. I just came back to this today and still looking at how to do this.

**Comment by RitterKnightCreative** [comment]
Here's the [code](https://github.com/RitterKnightCreative/craft-apple-pay/blob/master/_mobile-pay.html) that works for us.

Our Stripe  ``method.id`` is 2 but you'll want to change that to what you've set the Stripe gateway if you have more than one payment method. 

This also supports taxes and shipping so when you change your address, it'll fire back with the available shipping methods. After we get the City and State, the cart will get updated but the browser won't give you back the full address until the order is submitted which is kind of a pain.

Commerce wants to see First Name and Last Name as separate fields but AFAIK, that's not possible in V3 of Stripe (it was in V2) so we do a quick split, probably because Google Pay/Chrome autofill in chrome only shows Full Name. That function could probably be improved.

**Comment by philipboomy** [comment]
@RitterKnightCreative Thank you so much for sharing the code and the tips.

**Comment by aryehischechter** [comment]
@andris-sevcenko is there any update with progress on this being completed in this plugin?

**Comment by iamkeir** [comment]
Did this ever get solved? Either natively or as a plugin?

**Comment by Gs-Joel** [comment]
Bumping this again.
We have a client who is vey keen on using Google Pay.

**Comment by jamiematrix** [comment]
Also bumping. Have a client who just signed up for Stripe and wanting Apple and Google Pay.

**Reply by jamiematrix** [reply]
Also to note, am I right in thinking this gateway is using [Card Element from Stripe](https://stripe.com/docs/payments/elements)? If so, it's marked as Legacy - does that suggest in time it will be dropped?
